controlpanel/api/aws.py

```diff
@@ -24,248 +22,415 @@ def iam_arn(resource, account=settings.AWS_ACCOUNT_ID):
+    try:
+        iam.create_role(
+            RoleName=user.iam_role_name,
+            AssumeRolePolicyDocument=json.dumps(policy),
+        )
+    except iam.meta.client.exceptions.EntityAlreadyExistsException:
+        log.warning(f'Skipping creating Role {user.iam_role_name}: Already exists')
 
-    def _do(self, service, method, **kwargs):
-        if self.enabled:
-            client = self.client(service)
-            return getattr(client, method)(**kwargs)
+    role = iam.Role(user.iam_role_name)
+    role.attach_policy(
+        PolicyArn=iam_arn(f"policy/{READ_INLINE_POLICIES}"),
+    )
 
-    def create_bucket(self, name, region, acl="private"):
-        return self._do(
-            "s3",
-            "create_bucket",
-            Bucket=name,
-            ACL=acl,
-            CreateBucketConfiguration={"LocationConstraint": region},
-        )
 
-    def put_bucket_encryption(self, name):
-        self._do(
-            "s3",
-            "put_bucket_encryption",
-            Bucket=name,
-            ServerSideEncryptionConfiguration={
-                "Rules": [
-                    {"ApplyServerSideEncryptionByDefault": {"SSEAlgorithm": "AES256"}}
-                ]
-            },
-        )
+def delete_role(name):
+    """Delete the given IAM role and all inline policies"""
+    try:
+        role = boto3.resource('iam').Role(name)
+        role.load()
+    except role.meta.client.exceptions.NoSuchEntityException:
+        log.warning(f'Skipping delete of Role {name}: Does not exist')
+        return
 
-    def put_bucket_logging(self, name, target_bucket, target_prefix):
-        self._do(
-            "s3",
-            "put_bucket_logging",
-            Bucket=name,
-            BucketLoggingStatus={
-                "LoggingEnabled": {
-                    "TargetBucket": target_bucket,
-                    "TargetPrefix": target_prefix,
-                }
-            },
-        )
+    for policy in role.attached_policies.all():
+        role.detach_policy(PolicyArn=policy.arn)
 
-    def put_bucket_tagging(self, name, tags):
-        """Put bucket tagging
-
-        :param name: Bucket name
-        :param tags: Tags {key: value} e.g. {'buckettype': 'datawarehouse'}
-        :type name: str
-        :type tags: dict
-        """
-        self._do(
-            "s3",
-            "put_bucket_tagging",
-            Bucket=name,
-            Tagging={
-                "TagSet": [{"Key": str(k), "Value": str(v)} for k, v in tags.items()]
-            },
-        )
+    for policy in role.policies.all():
+        policy.delete()
+
+    role.delete()
 
-    def put_public_access_block(
-        self,
-        bucket_name,
-        block_public_acls=True,
-        ignore_public_acls=True,
-        block_public_policy=True,
-        restrict_public_buckets=True,
-    ):
-        self._do(
-            "s3",
-            "put_public_access_block",
+
+def create_bucket(bucket_name, is_data_warehouse=False):
```

Comment:
I'm not suggesting moving more things around as this was probably already there...

...but it feels strangely hi-level to have this `is_data_warehouse` argument in this function.
If the `aws.create_bucket()` function responsibility is to abstract the creation of an S3 bucket maybe it should receive a tags dictionary (and convert to the list of `{"Key": "...", "Value": "..."}`) instead of whether the S3 bucket is a data warehouse.

It makes me also wonder if this could belong in a `cluster.S3Bucket` class wrapping the `S3Bucket` model.
I notice that we also have a `cluster.create_bucket()` which is just calling this 🤷‍♂ 

Is `aws.create_bucket()` used somewhere else?

Maybe *long-term* if we really want to, we could move most of these AWS functions in the relevant methods in the `cluster` module and maybe just leave the class dealing with the IAM policy here...but that's for another refactoring.

Reply:
I had a look and it seems like `aws.create_bucket()` is only used in `cluster.Bucket.create()` which you just added so maybe we can bite the bullet and just move this logic there, but I'd still leave for a future refactoring.

Reply:
This `aws` module, is code for the platform to interact with AWS, so it definitely makes sense to have platform domain objects (eg data warehouse) referenced here. If we only presented a low-level interface to AWS here, there would be no need for this module - calling code could use `boto3` directly.

I think you are right that `cluster.create_bucket` should be moved into `cluster.Bucket.create`

Reply:
We don't have `cluster.create_bucket` - maybe that was a typo.